### PR TITLE
[WIP] r/aws_docdb_cluster_instance Add enable_performance_insights & performance_insights_kms_key_id argument

### PR DESCRIPTION
**Pull request by meetreks**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTS=TestAccXXX PKG=ec2

...
```

### CODE REVIEW

File: internal/service/docdb/cluster_instance.go [hunk 1]
```diff
@@ -188,12 +200,14 @@ func resourceClusterInstanceCreate(d *schema.ResourceData, meta interface{}) err
...
-		AutoMinorVersionUpgrade: aws.Bool(d.Get("auto_minor_version_upgrade").(bool)),
-		Tags:                    Tags(tags.IgnoreAWS()),
+		DBInstanceClass:             aws.String(d.Get("instance_class").(string)),
+		DBClusterIdentifier:         aws.String(d.Get("cluster_identifier").(string)),
+		Engine:                      aws.String(d.Get("engine").(string)),
+		PromotionTier:               aws.Int64(int64(d.Get("promotion_tier").(int))),
+		AutoMinorVersionUpgrade:     aws.Bool(d.Get("auto_minor_version_upgrade").(bool)),
+		Tags:                        Tags(tags.IgnoreAWS()),
+		EnablePerformanceInsights:   aws.Bool(d.Get("enable_performance_insights").(bool)),
+		PerformanceInsightsKMSKeyId: aws.String(d.Get("performance_insights_kms_key_id").(string)),
```
Comment (drewmullen) [inline internal/service/docdb/cluster_instance.go:210]: setting `PerformanceInsightsKMSKeyId` inside the `Input` object means the key is always passed. we only want to pass this key if its not null. instead this should be conditionally added to `createOpts`

File: internal/service/docdb/cluster_instance.go [hunk 2]
```diff
@@ -313,11 +327,13 @@ func resourceClusterInstanceRead(d *schema.ResourceData, meta interface{}) error
 	d.Set("availability_zone", db.AvailabilityZone)
 	d.Set("cluster_identifier", db.DBClusterIdentifier)
 	d.Set("dbi_resource_id", db.DbiResourceId)
+	//d.Set("enable_performance_insights", db.EnablePerformanceInsights)
```
Comment (drewmullen) [inline internal/service/docdb/cluster_instance.go:330]: Comment above that aws api does not expose these values and they should be uncommented if theyre ever added to `DescribeDBClusters` output

**Review by drewmullen — CHANGES_REQUESTED** [review]
Good start. Heres a few things: 
1. create a new file `.changelog/24948.txt` that contains your changelog entry. check the tf docs in this repo for how to format the file
2. update the resource documentation for these new parameters. make sure you include information about the defaults 
3. your update function needs to account for if these parameters are updated on an existing instance. what is sad about this is that since you cannot read these values via `describe` call, you are not able to detect drift. 
4. you need to run `terrafmt` on the test file to properly format the embedded hcl. here is an example for a different service/test file:
```bash
find ./internal/service/networkmanager -type f -name 'core_network_policy_document_data_source_test.go' \
    | sort -u \
    | xargs -I {} terrafmt fmt --fmtcompat {}
```
5. update the PR title to `[WIP] r/aws_docdb_cluster_instance Add `enable_performance_insights` & `performance_insights_kms_key_id` argument
6. add your acc test run output as a comment after your final commit is made
7. update the pr description to reflect any issues this PR closes

**Comment by meetreks** [comment]
Acceptance Tests
image

File: internal/service/docdb/cluster_instance.go [hunk 3]
```diff
@@ -311,11 +328,17 @@ func resourceClusterInstanceRead(d *schema.ResourceData, meta interface{}) error
 	d.Set("availability_zone", db.AvailabilityZone)
 	d.Set("cluster_identifier", db.DBClusterIdentifier)
 	d.Set("dbi_resource_id", db.DbiResourceId)
+	//Commented the below code as aws api does not expose these values and should be uncommented
+	//as soon as the vlues are available in the DescribeDBClusters output
+	//d.Set("enable_performance_insights", db.EnablePerformanceInsights)
```
Comment (daftkid) [inline internal/service/docdb/cluster_instance.go:333]: Won't it cause an infinite change of these two attributes during plan and apply? You cannot update these fields in Reading function, so how will the state get these values then?

File: internal/service/docdb/cluster_instance_test.go [hunk 4]
```diff
@@ -323,6 +358,53 @@ resource "aws_docdb_cluster_instance" "cluster_instances" {
...
+  preferred_instance_classes = ["db.t3.medium", "db.r4.large", "db.r5.large", "db.r5.xlarge"]
+}
+
+resource "aws_docdb_cluster_instance" "test" {
+  identifier         				= %[1]q
+  cluster_identifier 				= aws_docdb_cluster.default.id
+  instance_class     				= data.aws_docdb_orderable_db_instance.test.instance_class
+  promotion_tier     				= "3"
+  enable_performance_insights 		= true
+  performance_insights_kms_key_id 	= aws_kms_key.test.arn
```
Comment (daftkid) [inline internal/service/docdb/cluster_instance_test.go:403]: The format of TF code does not conform to standards. Could you please align all of the values or add newline before your two attributes?

**Comment by daftkid** [comment]
May I ask to include the initial issue reference that this PR is intended to close?
I mean to update this string "Relates OR Closes #0000" in the PR description

File: internal/service/docdb/cluster_instance.go (hunk 3 above)
Comment (meetreks) [reply internal/service/docdb/cluster_instance.go:333]: Only when the API gets updated, we can update the read finction with the new values.

File: internal/service/docdb/cluster_instance_test.go (hunk 4 above)
Comment (meetreks) [reply internal/service/docdb/cluster_instance_test.go:403]: Done as requested.

**Comment by drornir** [comment]
This PR seems abandoned by the original author, is there any way we can move this forward?
@daftkid here is the link to the original issue (I believe): https://github.com/hashicorp/terraform-provider-aws/issues/24857

**Pull request closed without merging**